test/gtest/ucs/test_event_set.cc

```diff
@@ -74,129 +186,89 @@ static void event_set_func3(void *callback_data, int events, void *arg)
-    ucs_event_set_cleanup(event_set);
+    thread_barrier();
 
-    pthread_join(tid, NULL);
+    event_set_wait(0u, 0, event_set_func3, NULL);
 
-    close(pipefd[0]);
-    close(pipefd[1]);
+    event_set_ctl(EVENT_SET_OP_DEL, m_pipefd[0], 0);
+    event_set_cleanup();
 }
 
-UCS_TEST_F(test_event_set, ucs_event_set_tmo_thread) {
-    ucs_sys_event_set_t *event_set = NULL;
-    pthread_t tid;
-    int ret;
-    int pipefd[2];
-    ucs_status_t status;
-    unsigned nread;
-
-    if (pipe(pipefd) == -1) {
-        UCS_TEST_MESSAGE << strerror(errno);
-        throw ucs::test_abort_exception();
-    }
+UCS_TEST_P(test_event_set, ucs_event_set_trig_modes) {
+    void *arg[] = { (void*)UCS_EVENT_SET_EXTRA_STRING,
+                    (void*)&UCS_EVENT_SET_EXTRA_NUM };
 
-    ret = pthread_barrier_init(&barrier, NULL, 2);
-    EXPECT_EQ(0, ret);
+    event_set_init(event_set_read_func);
+    event_set_ctl(EVENT_SET_OP_ADD, m_pipefd[0],
+                  UCS_EVENT_SET_EVREAD);
 
-    ret = pthread_create(&tid, NULL, event_set_tmo_func, (void *)&pipefd);
-    if (ret) {
-        UCS_TEST_MESSAGE << strerror(errno);
-        throw ucs::test_abort_exception();
-    }
+    thread_barrier();
 
-    status = ucs_event_set_create(&event_set);
-    EXPECT_EQ(UCS_OK, status);
-    EXPECT_TRUE(event_set != NULL);
+    /* Test level-triggered mode (default) */
+    for (int i = 0; i < 10; i++) {
+        event_set_wait(1u, 0, event_set_func4, NULL);
+    }
 
-    status = ucs_event_set_add(event_set, pipefd[0], UCS_EVENT_SET_EVREAD,
-                               NULL);
-    EXPECT_EQ(UCS_OK,status);
+    /* Test edge-triggered mode */
+    /* Set edge-triggered mode */
+    event_set_ctl(EVENT_SET_OP_MOD, m_pipefd[0],
+                  (ucs_event_set_type_t)(UCS_EVENT_SET_EVREAD |
+                                         UCS_EVENT_SET_EVET));
 
-    ret = pthread_barrier_wait(&barrier);
-    EXPECT_TRUE((ret == 0) || (ret == PTHREAD_BARRIER_SERIAL_THREAD));
+    /* Should have only one event to read */
+    event_set_wait(1u, 0, event_set_func4, NULL);
 
-    pthread_join(tid, NULL);
-    pthread_barrier_destroy(&barrier);
+    /* Should not read nothing */
+    for (int i = 0; i < 10; i++) {
+        event_set_wait(0u, 0, event_set_func1, arg);
+    }
 
-    /* Check for events on pipe fd */
-    nread  = ucs_sys_event_set_max_wait_events;
-    status = ucs_event_set_wait(event_set, &nread, 0, event_set_func3, NULL);
-    EXPECT_EQ(0u, nread);
-    EXPECT_EQ(UCS_OK, status);
-    ucs_event_set_cleanup(event_set);
+    /* Call the function below directly to read all outstanding data from pipe fd */
+    event_set_func1((void*)(uintptr_t)m_pipefd[0], UCS_EVENT_SET_EVREAD, arg);
 
-    close(pipefd[0]);
-    close(pipefd[1]);
+    event_set_ctl(EVENT_SET_OP_DEL, m_pipefd[0], 0);
+    event_set_cleanup();
 }
+
+INSTANTIATE_TEST_CASE_P(ext_fd, test_event_set,
+                        ::testing::Values(static_cast<int>(UCS_EVENT_SET_EXTERNAL_FD)));
```

Comment:
minor: over 80 columns. (But I'm not sure how to format this line)

Reply:
fixed